### ISSUE TITLE
Consider monorepo

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#1599]
e.g. lerna, Yarn workspaces

The first use case for monorepo structure would be the validator (#1352). For example we might have three packages within the repo:
- `renovate`
- `@renovate/validate`
- `@renovate/lint`

Where `@renovate/validate` is a library shared by both renovate and `@renovate/lint`, and `@renovate/lint` is a CLI tool that can be added to tests.

Or perhaps `@renovate/validate` could be both a library (required by renovate) as well as the CLI tool in one.

Either way, there would be a benefit in splitting into a monorepo.

lerna seems to be the main option, and Yarn Workspaces provides the best way to optimise installations.

The big issue would be versioning and publishing:
- version numbering and synchronisation
- semantic release

For version numbering, lerna offers the option of synchronised or independent versions. Although it makes sense to keep validation/lint/cli all synchronised to the main renovate version, there may be packages in future that are better to keep independently versioned. eg some of our JSON, npm, or docker utilities that could be used by a wider audience.

For semantic release.. the main issue is how to keep up automatic releasing. If we have independent versioning, I think that might be impossible with today's tools. Additionally:
- It makes sense if the validator is the same version as the renovate tool itself, but
- It would be annoying to keep bumping the validator releases unnecessarily, especially if a lot of people have it in their `devDependencies`

**Comment by rarkins** [comment]
BTW `renovate` is currently huge (~300MB installed) mainly due to it including `npm`, `yarn` and `pnpm` as `dependencies`. So if the validator script is a standalone module that "requires" `renovate` and then just uses its validator module then the size of the validator would be unnecessarily large unless it somehow treeshakes it.

**Comment by destroyerofbuilds** [comment]
> Either way, there would be a benefit in splitting into a monorepo.

We use the same validation technique created by the `seek` configuration package and documented in #1352. It, obviously, dependents on the entire `renovate` package, which definitely adds install time overhead. However, that validation technique has saved us more than once, so the overhead as been well worth it.

> lerna seems to be the main option

Lerna seems to be a pretty good option. I just started working with the tool as part of [my work on `conventional-changelog`](https://github.com/conventional-changelog/conventional-changelog/commit/c5f993a438b5430928e24bd2fc2b817e953a30ff).

> The big issue would be versioning and publishing:

Lerna uses `conventional-recommended-bump` under the hood. The same as `semantic-release`.

You can get roughly the same behavior as `semantic-release` with something like:

```bash
lerna publish --conventional-commits --yes
```

Though, keep in mind that Lerna creates a git commit with updated `package.json` files (It changes the `version` field), and updated `CHANGELOG.md` files. That is different than the default settings in `semantic-release`.

**Comment by destroyerofbuilds** [comment]
> It makes sense if the validator is the same version as the renovate tool itself

So here's an issue we ran into: Our `renovate-config` package depended on a Renovate version earlier than version `11.27.0`, yet our Renovate service was deployed with a version greater than [`11.17.0`](https://github.com/renovateapp/renovate/releases/tag/v11.27.0) (which contains the `allowedVersions` feature).

When we updated our `renovate-config` package with `allowedVersions`, our tests failed, at least until we realized that we needed to upgrade our lockfile.

Perhaps that should be obvious, but it's a little weird that the Renovate service could be running a different version than what is being used to lint/validate the configuration package.
